types/etc.go

```diff
@@ -121,7 +121,8 @@ func IsNonBinaryStr(ft *FieldType) bool {
 func NeedRestoredData(ft *FieldType) bool {
 	if collate.NewCollationEnabled() &&
 		IsNonBinaryStr(ft) &&
-		!(collate.IsBinCollation(ft.GetCollate()) && !IsTypeVarchar(ft.GetType())) {
+		(!collate.IsBinCollation(ft.GetCollate()) || IsTypeVarchar(ft.GetType())) &&
+		ft.GetCollate() != "utf8mb4_0900_bin" {
```

Comment:
We have added "utf8mb4_0900_bin" to "IsBinCollation" in this PR, so could we remove this line?

Reply:
On the other hand: If `NeedRestoredData` is true, then:

1. If `IsBinCollation` is true, the count of tailing spaces will be written into the value as an integer.
2. If `IsBinCollation` is false, the original string will be written into the value as a string.